Clean up temp files from installer download

ryfu-msft opened the pull request:
The changes in this pull request ensure that the temp files from the downloaded installer are deleted before the end of the command execution. This should fix #40, where the installer hash does not match due to old installer files not being cleaned up from the %TEMP% directory.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-create/pull/56)

palenshus commented:
Looks like we have a bug, in DownloadFileAsync, if the file already exists, but with a different length, we download the new one, but looks like that doesn't replace the old one, it just writes overwrites it. Meaning, if the previous one is larger, we'll download a smaller one and write those bytes over the old one, creating a bogus file.

The fix is simple, and I think sufficient to protect us here, add a File.Delete:

            if (!File.Exists(targetFile) || new FileInfo(targetFile).Length != downloadSize)
            {
                File.Delete(targetFile);
                using var targetFileStream = File.OpenWrite(targetFile);

What do you think?

palenshus commented on src/WingetCreateCLI/Commands/NewCommand.cs:
```diff
@@ -205,8 +207,9 @@ private static void PromptRequiredProperties<T>(T manifest, VersionManifest vers
+                    property.SetValue(manifest, result);
+                    string formattedResult = result.ToString().Replace("{", string.Empty).Replace("}", string.Empty);
+                    Logger.Trace($"Property [{property.Name}] set to the value [{formattedResult}]");
                 }
```
I don't think this addresses their issues. All the stack traces seem to originate in this call in Program.cs:

    Logger.Error(ex.ToString());

I believe that what's possibly happening is that the exception method has a format string inside of it, for example "{version}". You can repro with:

wingetcreate new [https://foo.com/{version}.exe](https://foo.com/{version}.exe)

palenshus replied:
We should fix this in a separate PR

ryfu-msft replied:
I reverted my changes because your recent PR fixes this issue.

palenshus replied:
(Fixed in separate PR)

ryfu-msft commented:
> Looks like we have a bug, in DownloadFileAsync, if the file already exists, but with a different length, we download the new one, but looks like that doesn't replace the old one, it just writes overwrites it. Meaning, if the previous one is larger, we'll download a smaller one and write those bytes over the old one, creating a bogus file.
> 
> The fix is simple, and I think sufficient to protect us here, add a File.Delete:
> 
> ```
>         if (!File.Exists(targetFile) || new FileInfo(targetFile).Length != downloadSize)
>         {
>             File.Delete(targetFile);
>             using var targetFileStream = File.OpenWrite(targetFile);
> ```
> 
> What do you think?

100% agree. This is a much better solution. I added your suggested changes and removed mine

palenshus commented on NOTICE.txt:
```diff
@@ -369,18 +369,17 @@ Permission is hereby granted, free of charge, to any person obtaining a copy of
+
+---------------------------------------------------------
+
+jose-jwt 3.1.1 - MIT
```
>jose [](http://example.com/codeflow?start=0&length=4)

Thanks!

palenshus commented:
>     using CommandLine;

Can you reset this file since it's no longer being changed?

---
Refers to: src/WingetCreateCLI/Commands/NewCommand.cs:13 in 9c904ab. [](commit_id = 9c904abf36320082d493027a0a49762dfccccb27, deletion_comment = False)

palenshus reviewed (approved):
:shipit:
